Automattic/sensei
# Frontend for progress bar related to pagination

Fixes #4542

### Changes proposed in this Pull Request
- Reuse ProgressBar component to display quiz progress for user
- How is it done: 
1. On page load with before quiz hook element is added with id _'progress-bar-quiz'_
2. When quiz data is loaded if **progress bar** **is enabled** replace _'progress-bar-quiz'_ with ProgressBar react component
3. Pass the data to the component: 
- How many questions is there in the quiz
- How many questions user answered
- And from metadata: height, radius, color, and backgroundColor for the bar 
 
### Testing instructions
Enable the feature flag: `add_filter( 'sensei_feature_flag_quiz_pagination', '__return_true' );`
Whatever is changed Lesson needs to be UPDATED (click update button) in order to see the changes 

#### 1. Hide Settings for progress bar if 'Multi page' (pagination) is not selected and progress bar settings displayed when 'Multi-page' option selected in pagination: 

| Pagination disabled  | Pagination enabled | Colors Settings (this isn't custom property but form wordpress library) |
| -- | -- | -- |
| ![image](https://user-images.githubusercontent.com/7208249/148780765-46689e8d-957f-4123-a98a-206c0e287dfe.png) | ![image](https://user-images.githubusercontent.com/7208249/148780859-dfe090e1-ba23-40f5-9374-e04cab8d83bf.png) | ![image](https://user-images.githubusercontent.com/7208249/148780896-76693237-cb9a-4a02-b361-48d8b90bfba2.png) |

#### 2. Progress bar not displayed in case pagination is not enabled, progress bar is not displayed if progress bar toogle is not enabled  
| Progress bar disabled  | Pagination disabled |
| -- | -- |  
| ![image](https://user-images.githubusercontent.com/7208249/148781005-126a4c70-1134-4b73-97db-c35f989b1a19.png) | ![image](https://user-images.githubusercontent.com/7208249/148781039-66fb9f01-f03a-43e4-b585-2351d4d5e9d5.png) |

#### 3. Progress bar enabled selected and progress bar visible above the questions. If no configuration is set, use default 
| Pagination and progress bar enabled, settings not changes  | Display progress bar with default settings |
| -- | -- |  
| ![image](https://user-images.githubusercontent.com/7208249/148781298-eb871547-6c36-40ef-987b-e8a4fd4d67c7.png) | ![image](https://user-images.githubusercontent.com/7208249/148781319-735db6e8-e0b9-486c-91c7-e1569157c3c7.png) |

#### 4. Set custom width and height and check the changes on view.
| Custom width and height | Display progress bar with custom width and height |
| -- | -- |  
| ![image](https://user-images.githubusercontent.com/7208249/148781705-b2033e7f-8a1f-4a6d-b31f-a9c4dd6ca720.png) | ![image](https://user-images.githubusercontent.com/7208249/148781752-80765a02-c5b2-4ef7-b039-e7a19ab999bd.png) |

#### 5. Set the custom color and background color.
| Set custom color and background color | Display progress bar with custom color and background color |
| -- | -- |
| ![image](https://user-images.githubusercontent.com/7208249/148781929-ce109f47-d359-4028-a6d6-51d2b67c2f47.png) | ![image](https://user-images.githubusercontent.com/7208249/148781908-db049992-d072-4708-bc28-6ab8c2b660aa.png) |
 
### New/Updated Hooks

Added function ```the_quiz_progress_bar``` 
```add_action( 'sensei_single_quiz_questions_before', array( 'Sensei_Quiz', 'the_quiz_progress_bar' ), 20 );```
To the **sensei_single_quiz_questions_before** hook 
Adding element before the quiz so that it can be replaced with Progress Bar for pagination later on
